## System
- Hot Pepper Gourmet(hpg): 이용자들이 식당을 찾고 온라인 예약할 수 있는 시스템
- AirREGI / Restaurant Board(air): 예약 관리 및 계산 시스템

## air_reserve.csv
- air_store_id: 식당의 ID(air)
- visit_datetime: 예약 시간(방문 예정 시간)
- reserve_datetime: 예약 등록을 한 시간
- reserve_visitors: 예정 방문자 수

## hpg_reserve.csv
- hpg_store_id: 식당의 ID(hpg)
- visit_datetime: 예약 시간(방문 예정 시간)
- reserve_datetime: 예약 등록을 한 시간
- reserve_visitors: 예정 방문자 수

## air_store_info.csv
식당에서 air 시스템에 등록한 정보
- air_store_id: 식당 ID(air)
- air_genre_name: 식당의 종류(ex.이탈리안, 이자카야) 
- air_area_name: 지역
- latitude: 경도 (해당 지역의 위도, 경도)
- longitude: 위도

## hpg_store_info.csv
식당에서 hpg 시스템에 등록한 정보
- air_store_id: 식당 ID(hpg)
- air_genre_name: 식당의 종류(ex.이탈리안, 이자카야) 
- air_area_name: 지역
- latitude: 경도 (해당 지역의 위도, 경도)
- longitude: 위도

## store_id_relation.csv
두 시스템에서의 ID가 매칭되어 있다.
- hpg_store_id
- air_store_id

## air_visit_data.csv
방문 기록이 담겨있다.
- air_store_id: 식당의 ID(air)
- visit_date - 방문 날짜
- visitors - 방문자의 수

## date_info.csv
- calendar_date: 날짜(년, 월, 일)
- day_of_week: 요일
- holiday_flg - 일본 기준 휴일 여부(1,0)

## sample_submission.csv
제출 양식
- id: air 기준 식당 ID와 방문 날짜가 결합된 형태
- visiotors: id에 해당하는 날짜, 식당에 방문할 손님의 수 예측


# 데이터 불러오기 및 확인

In [7]:
import pandas as pd

air_reserve = pd.read_csv(r"C:\Users\crosstar\OneDrive - 고려대학교\Desktop\recruit-restaurant-visitor-forecasting\air_reserve.csv")
hpg_reserve = pd.read_csv(r"C:\Users\crosstar\OneDrive - 고려대학교\Desktop\recruit-restaurant-visitor-forecasting\hpg_reserve.csv")
air_store_info = pd.read_csv(r"C:\Users\crosstar\OneDrive - 고려대학교\Desktop\recruit-restaurant-visitor-forecasting\air_store_info.csv")
hpg_store_info = pd.read_csv(r"C:\Users\crosstar\OneDrive - 고려대학교\Desktop\recruit-restaurant-visitor-forecasting\hpg_store_info.csv")
store_id_relation = pd.read_csv(r"C:\Users\crosstar\OneDrive - 고려대학교\Desktop\recruit-restaurant-visitor-forecasting\store_id_relation.csv")
air_visit_data = pd.read_csv(r"C:\Users\crosstar\OneDrive - 고려대학교\Desktop\recruit-restaurant-visitor-forecasting\air_visit_data.csv")
date_info = pd.read_csv(r"C:\Users\crosstar\OneDrive - 고려대학교\Desktop\recruit-restaurant-visitor-forecasting\date_info.csv")
stores_to_pred = pd.read_csv(r"C:\Users\crosstar\OneDrive - 고려대학교\Desktop\recruit-restaurant-visitor-forecasting\sample_submission.csv")


## air_store_info

In [8]:
air_store_info.shape

(829, 5)

In [9]:
air_store_info.head()

air_store_id  air_genre_name                 air_area_name  \
0  air_0f0cdeee6c9bf3d7  Italian/French  Hyōgo-ken Kōbe-shi Kumoidōri   
1  air_7cc17a324ae5c7dc  Italian/French  Hyōgo-ken Kōbe-shi Kumoidōri   
2  air_fee8dcf4d619598e  Italian/French  Hyōgo-ken Kōbe-shi Kumoidōri   
3  air_a17f0778617c76e2  Italian/French  Hyōgo-ken Kōbe-shi Kumoidōri   
4  air_83db5aff8f50478e  Italian/French  Tōkyō-to Minato-ku Shibakōen   

    latitude   longitude  
0  34.695124  135.197853  
1  34.695124  135.197853  
2  34.695124  135.197853  
3  34.695124  135.197853  
4  35.658068  139.751599

In [10]:
# 중복값 확인
air_store_info['air_store_id'].duplicated().sum()

0

## hpg_store_info

In [11]:
hpg_store_info.shape

(4690, 5)

In [12]:
hpg_store_info.head()

hpg_store_id  hpg_genre_name                  hpg_area_name  \
0  hpg_6622b62385aec8bf  Japanese style  Tōkyō-to Setagaya-ku Taishidō   
1  hpg_e9e068dd49c5fa00  Japanese style  Tōkyō-to Setagaya-ku Taishidō   
2  hpg_2976f7acb4b3a3bc  Japanese style  Tōkyō-to Setagaya-ku Taishidō   
3  hpg_e51a522e098f024c  Japanese style  Tōkyō-to Setagaya-ku Taishidō   
4  hpg_e3d0e1519894f275  Japanese style  Tōkyō-to Setagaya-ku Taishidō   

    latitude   longitude  
0  35.643675  139.668221  
1  35.643675  139.668221  
2  35.643675  139.668221  
3  35.643675  139.668221  
4  35.643675  139.668221

In [13]:
# 중복값 확인
hpg_store_info['hpg_store_id'].duplicated().sum()

0

air 시스템에는 829개, hpg 시스템에는 4690개의 식당이 등록되어 있습니다.

## air_reserve

In [14]:
air_reserve.shape

(92378, 4)

In [15]:
air_reserve.head()

air_store_id       visit_datetime     reserve_datetime  \
0  air_877f79706adbfb06  2016-01-01 19:00:00  2016-01-01 16:00:00   
1  air_db4b38ebe7a7ceff  2016-01-01 19:00:00  2016-01-01 19:00:00   
2  air_db4b38ebe7a7ceff  2016-01-01 19:00:00  2016-01-01 19:00:00   
3  air_877f79706adbfb06  2016-01-01 20:00:00  2016-01-01 16:00:00   
4  air_db80363d35f10926  2016-01-01 20:00:00  2016-01-01 01:00:00   

   reserve_visitors  
0                 1  
1                 3  
2                 6  
3                 2  
4                 5

In [16]:
air_reserve.duplicated().sum()

2770

In [17]:
air_reserve[(air_reserve['air_store_id'] == 'air_789466e488705c93') & (air_reserve['visit_datetime'] == '2016-01-04 17:00:00') ]

air_store_id       visit_datetime     reserve_datetime  \
47  air_789466e488705c93  2016-01-04 17:00:00  2016-01-04 17:00:00   
48  air_789466e488705c93  2016-01-04 17:00:00  2016-01-04 17:00:00   

    reserve_visitors  
47                 2  
48                 2

In [18]:
air_reserve[(air_reserve['air_store_id'] == 'air_e55abd740f93ecc4') & (air_reserve['visit_datetime'] == '2016-01-05 20:00:00') ]

air_store_id       visit_datetime     reserve_datetime  \
91  air_e55abd740f93ecc4  2016-01-05 20:00:00  2016-01-05 18:00:00   
92  air_e55abd740f93ecc4  2016-01-05 20:00:00  2016-01-05 18:00:00   

    reserve_visitors  
91                 2  
92                 2

중복값이 많아 확인해본 결과 완벽하게 동일한 예약 내용들이 있었습니다. 취소 후 재예약을 한 것이 아닌가 예상됩니다. **다만, 재예약의 경우 reserve_datetime은 변할 수 있는데 이 부분을 어떻게 확인해야 될지는 더 생각해봐야할 부분입니다.**

In [19]:
air_reserve[['air_store_id','visit_datetime','reserve_visitors']].duplicated().sum()

14609

**reserve_datetime을 제외한 부분의 중복값들이 확실히 많기는 하나 재예약인지 다른 예약인지 구분할 수는 없습니다.
->solution 필요함**

## hpg_reserve

In [20]:
hpg_reserve.shape

(2000320, 4)

In [21]:
hpg_reserve.head()

hpg_store_id       visit_datetime     reserve_datetime  \
0  hpg_c63f6f42e088e50f  2016-01-01 11:00:00  2016-01-01 09:00:00   
1  hpg_dac72789163a3f47  2016-01-01 13:00:00  2016-01-01 06:00:00   
2  hpg_c8e24dcf51ca1eb5  2016-01-01 16:00:00  2016-01-01 14:00:00   
3  hpg_24bb207e5fd49d4a  2016-01-01 17:00:00  2016-01-01 11:00:00   
4  hpg_25291c542ebb3bc2  2016-01-01 17:00:00  2016-01-01 03:00:00   

   reserve_visitors  
0                 1  
1                 3  
2                 2  
3                 5  
4                13

In [22]:
hpg_reserve.duplicated().sum()

16279

hpg_reserve도 동일합니다.

## store_id_relation

In [23]:
store_id_relation.shape

(150, 2)

In [24]:
store_id_relation.head()

air_store_id          hpg_store_id
0  air_63b13c56b7201bd9  hpg_4bc649e72e2a239a
1  air_a24bf50c3e90d583  hpg_c34b496d0305a809
2  air_c7f78b4f3cba33ff  hpg_cd8ae0d9bbd58ff9
3  air_947eb2cae4f3e8f2  hpg_de24ea49dc25d6b8
4  air_965b2e0cf4119003  hpg_653238a84804d8e7

In [25]:
store_id_relation.duplicated().sum()

0

## air_visit_data

In [26]:
air_visit_data.shape

(252108, 3)

In [27]:
air_visit_data.head()

air_store_id  visit_date  visitors
0  air_ba937bf13d40fb24  2016-01-13        25
1  air_ba937bf13d40fb24  2016-01-14        32
2  air_ba937bf13d40fb24  2016-01-15        29
3  air_ba937bf13d40fb24  2016-01-16        22
4  air_ba937bf13d40fb24  2016-01-18         6

In [28]:
air_visit_data.duplicated().sum()

0

## date_info

In [30]:
date_info.shape

(517, 3)

In [31]:
date_info.head()

calendar_date day_of_week  holiday_flg
0    2016-01-01      Friday            1
1    2016-01-02    Saturday            1
2    2016-01-03      Sunday            1
3    2016-01-04      Monday            0
4    2016-01-05     Tuesday            0

## stores to pred

In [32]:
stores_to_pred.head()

id  visitors
0  air_00a91d42b08b08d9_2017-04-23         0
1  air_00a91d42b08b08d9_2017-04-24         0
2  air_00a91d42b08b08d9_2017-04-25         0
3  air_00a91d42b08b08d9_2017-04-26         0
4  air_00a91d42b08b08d9_2017-04-27         0

Split **id** -> **air_store_id** and **visit_date** 

In [33]:
id_split = stores_to_pred['id'].str.split('_', expand=True)
id_split.columns = ['source', 'store_id', 'visit_date']
stores_to_pred = pd.concat([stores_to_pred, id_split], axis=1)
stores_to_pred['air_store_id']=stores_to_pred['source'].str.cat(stores_to_pred['store_id'], sep='_')
stores_to_pred = stores_to_pred[['id', 'air_store_id', 'visit_date', 'visitors']]
stores_to_pred['visit_date'] = pd.to_datetime(stores_to_pred['visit_date'])

stores_to_pred.head()

id          air_store_id visit_date  visitors
0  air_00a91d42b08b08d9_2017-04-23  air_00a91d42b08b08d9 2017-04-23         0
1  air_00a91d42b08b08d9_2017-04-24  air_00a91d42b08b08d9 2017-04-24         0
2  air_00a91d42b08b08d9_2017-04-25  air_00a91d42b08b08d9 2017-04-25         0
3  air_00a91d42b08b08d9_2017-04-26  air_00a91d42b08b08d9 2017-04-26         0
4  air_00a91d42b08b08d9_2017-04-27  air_00a91d42b08b08d9 2017-04-27         0

In [34]:
stores_to_pred.shape

(32019, 4)

In [35]:
stores_to_pred.air_store_id.nunique()

821

In [36]:
stores_to_pred.visit_date.nunique()

39

## Merge Full

air => **air_reserve + air_store_info** + store_id_relation

hpg => **hpg_reserve + hpg_store_info** + store_id_relation

In [41]:
air = pd.merge(air_reserve, air_store_info, on='air_store_id')
hpg = pd.merge(hpg_reserve, hpg_store_info, on='hpg_store_id')

airrel = pd.merge(air, store_id_relation, how='left', on='air_store_id')
hpgrel = pd.merge(hpg, store_id_relation, how='left', on='hpg_store_id')
full = pd.merge(airrel, hpgrel, how='outer')

print("")
print("there are ",len(air)," restaurants with AIR and ",len(hpg)," with HPG.",len(store_id_relation),' have both.')

there are  92378  restaurants with AIR and  871308  with HPG. 150  have both.


In [42]:
full

air_store_id       visit_datetime     reserve_datetime  \
0       air_877f79706adbfb06  2016-01-01 19:00:00  2016-01-01 16:00:00   
1       air_877f79706adbfb06  2016-01-01 20:00:00  2016-01-01 16:00:00   
2       air_877f79706adbfb06  2016-01-02 18:00:00  2016-01-01 16:00:00   
3       air_877f79706adbfb06  2016-01-02 21:00:00  2016-01-01 16:00:00   
4       air_877f79706adbfb06  2016-01-08 21:00:00  2016-01-04 20:00:00   
...                      ...                  ...                  ...   
963681                   NaN  2017-04-24 18:00:00  2017-04-20 14:00:00   
963682                   NaN  2017-04-24 19:00:00  2017-04-22 14:00:00   
963683                   NaN  2017-04-24 19:00:00  2017-04-22 14:00:00   
963684                   NaN  2017-04-24 19:00:00  2017-04-10 17:00:00   
963685                   NaN  2017-04-28 18:00:00  2017-04-10 17:00:00   

        reserve_visitors air_genre_name                 air_area_name  \
0                      1  Japanese food  Tōkyō-to Minato-ku Shibakōen   
1                      2  Japanese food  Tōkyō-to Minato-ku Shibakōen   
2                      2  Japanese food  Tōkyō-to Minato-ku Shibakōen   
3                      2  Japanese food  Tōkyō-to Minato-ku Shibakōen   
4                      2  Japanese food  Tōkyō-to Minato-ku Shibakōen   
...                  ...            ...                           ...   
963681                 8            NaN                           NaN   
963682                14            NaN                           NaN   
963683                15            NaN                           NaN   
963684                18            NaN                           NaN   
963685                22            NaN                           NaN   

         latitude   longitude          hpg_store_id  hpg_genre_name  \
0       35.658068  139.751599                   NaN             NaN   
1       35.658068  139.751599                   NaN             NaN   
2       35.658068  139.751599                   NaN             NaN   
3       35.658068  139.751599                   NaN             NaN   
4       35.658068  139.751599                   NaN             NaN   
...           ...         ...                   ...             ...   
963681  35.682504  139.773619  hpg_e106ddef2b137959  Japanese style   
963682  35.682504  139.773619  hpg_e106ddef2b137959  Japanese style   
963683  35.682504  139.773619  hpg_e106ddef2b137959  Japanese style   
963684  35.682504  139.773619  hpg_e106ddef2b137959  Japanese style   
963685  35.682504  139.773619  hpg_e106ddef2b137959  Japanese style   

                hpg_area_name  
0                         NaN  
1                         NaN  
2                         NaN  
3                         NaN  
4                         NaN  
...                       ...  
963681  Tōkyō-to Chūō-ku None  
963682  Tōkyō-to Chūō-ku None  
963683  Tōkyō-to Chūō-ku None  
963684  Tōkyō-to Chūō-ku None  
963685  Tōkyō-to Chūō-ku None  

[963686 rows x 11 columns]

In [65]:
# datetime을 날짜와 시간으로 분리
full['visit_date'] = full['visit_datetime'].map(lambda x: x.split()[0])
full['visit_time'] = full['visit_datetime'].map(lambda x: x.split()[1])
full.drop(columns=['visit_datetime'], inplace=True)

In [71]:
full

air_store_id     reserve_datetime  reserve_visitors  \
0       air_877f79706adbfb06  2016-01-01 16:00:00                 1   
1       air_877f79706adbfb06  2016-01-01 16:00:00                 2   
2       air_877f79706adbfb06  2016-01-01 16:00:00                 2   
3       air_877f79706adbfb06  2016-01-01 16:00:00                 2   
4       air_877f79706adbfb06  2016-01-04 20:00:00                 2   
...                      ...                  ...               ...   
963681                   NaN  2017-04-20 14:00:00                 8   
963682                   NaN  2017-04-22 14:00:00                14   
963683                   NaN  2017-04-22 14:00:00                15   
963684                   NaN  2017-04-10 17:00:00                18   
963685                   NaN  2017-04-10 17:00:00                22   

       air_genre_name                 air_area_name   latitude   longitude  \
0       Japanese food  Tōkyō-to Minato-ku Shibakōen  35.658068  139.751599   
1       Japanese food  Tōkyō-to Minato-ku Shibakōen  35.658068  139.751599   
2       Japanese food  Tōkyō-to Minato-ku Shibakōen  35.658068  139.751599   
3       Japanese food  Tōkyō-to Minato-ku Shibakōen  35.658068  139.751599   
4       Japanese food  Tōkyō-to Minato-ku Shibakōen  35.658068  139.751599   
...               ...                           ...        ...         ...   
963681            NaN                           NaN  35.682504  139.773619   
963682            NaN                           NaN  35.682504  139.773619   
963683            NaN                           NaN  35.682504  139.773619   
963684            NaN                           NaN  35.682504  139.773619   
963685            NaN                           NaN  35.682504  139.773619   

                hpg_store_id  hpg_genre_name          hpg_area_name  \
0                        NaN             NaN                    NaN   
1                        NaN             NaN                    NaN   
2                        NaN             NaN                    NaN   
3                        NaN             NaN                    NaN   
4                        NaN             NaN                    NaN   
...                      ...             ...                    ...   
963681  hpg_e106ddef2b137959  Japanese style  Tōkyō-to Chūō-ku None   
963682  hpg_e106ddef2b137959  Japanese style  Tōkyō-to Chūō-ku None   
963683  hpg_e106ddef2b137959  Japanese style  Tōkyō-to Chūō-ku None   
963684  hpg_e106ddef2b137959  Japanese style  Tōkyō-to Chūō-ku None   
963685  hpg_e106ddef2b137959  Japanese style  Tōkyō-to Chūō-ku None   

        visit_date visit_time  
0       2016-01-01   19:00:00  
1       2016-01-01   20:00:00  
2       2016-01-02   18:00:00  
3       2016-01-02   21:00:00  
4       2016-01-08   21:00:00  
...            ...        ...  
963681  2017-04-24   18:00:00  
963682  2017-04-24   19:00:00  
963683  2017-04-24   19:00:00  
963684  2017-04-24   19:00:00  
963685  2017-04-28   18:00:00  

[963686 rows x 12 columns]

In [72]:
# 요일과 휴일 여부가 담긴 date_info와 merge
full_date = pd.merge(full, date_info, how='left', left_on='visit_date', right_on='calendar_date')
full_date.drop(columns=['calendar_date'], inplace=True)

In [73]:
full_date

air_store_id     reserve_datetime  reserve_visitors  \
0       air_877f79706adbfb06  2016-01-01 16:00:00                 1   
1       air_877f79706adbfb06  2016-01-01 16:00:00                 2   
2       air_877f79706adbfb06  2016-01-01 16:00:00                 2   
3       air_877f79706adbfb06  2016-01-01 16:00:00                 2   
4       air_877f79706adbfb06  2016-01-04 20:00:00                 2   
...                      ...                  ...               ...   
963681                   NaN  2017-04-20 14:00:00                 8   
963682                   NaN  2017-04-22 14:00:00                14   
963683                   NaN  2017-04-22 14:00:00                15   
963684                   NaN  2017-04-10 17:00:00                18   
963685                   NaN  2017-04-10 17:00:00                22   

       air_genre_name                 air_area_name   latitude   longitude  \
0       Japanese food  Tōkyō-to Minato-ku Shibakōen  35.658068  139.751599   
1       Japanese food  Tōkyō-to Minato-ku Shibakōen  35.658068  139.751599   
2       Japanese food  Tōkyō-to Minato-ku Shibakōen  35.658068  139.751599   
3       Japanese food  Tōkyō-to Minato-ku Shibakōen  35.658068  139.751599   
4       Japanese food  Tōkyō-to Minato-ku Shibakōen  35.658068  139.751599   
...               ...                           ...        ...         ...   
963681            NaN                           NaN  35.682504  139.773619   
963682            NaN                           NaN  35.682504  139.773619   
963683            NaN                           NaN  35.682504  139.773619   
963684            NaN                           NaN  35.682504  139.773619   
963685            NaN                           NaN  35.682504  139.773619   

                hpg_store_id  hpg_genre_name          hpg_area_name  \
0                        NaN             NaN                    NaN   
1                        NaN             NaN                    NaN   
2                        NaN             NaN                    NaN   
3                        NaN             NaN                    NaN   
4                        NaN             NaN                    NaN   
...                      ...             ...                    ...   
963681  hpg_e106ddef2b137959  Japanese style  Tōkyō-to Chūō-ku None   
963682  hpg_e106ddef2b137959  Japanese style  Tōkyō-to Chūō-ku None   
963683  hpg_e106ddef2b137959  Japanese style  Tōkyō-to Chūō-ku None   
963684  hpg_e106ddef2b137959  Japanese style  Tōkyō-to Chūō-ku None   
963685  hpg_e106ddef2b137959  Japanese style  Tōkyō-to Chūō-ku None   

        visit_date visit_time day_of_week  holiday_flg  
0       2016-01-01   19:00:00      Friday            1  
1       2016-01-01   20:00:00      Friday            1  
2       2016-01-02   18:00:00    Saturday            1  
3       2016-01-02   21:00:00    Saturday            1  
4       2016-01-08   21:00:00      Friday            0  
...            ...        ...         ...          ...  
963681  2017-04-24   18:00:00      Monday            0  
963682  2017-04-24   19:00:00      Monday            0  
963683  2017-04-24   19:00:00      Monday            0  
963684  2017-04-24   19:00:00      Monday            0  
963685  2017-04-28   18:00:00      Friday            0  

[963686 rows x 14 columns]